src/database/import.ts

```diff
@@ -190,3 +213,63 @@ export default class DatabaseImporter {
     return [root, key].join("/").replace("//", "/");
   }
 }
+
+/**
+ * Batches chunked JSON data up to the specified byte size limit. The emitted batch is an object
+ * containing chunks keyed by pathname.
+ */
+class BatchChunks extends stream.Transform {
+  private batch: Data = { json: {}, pathname: "" };
+  private size = 0;
+
+  constructor(private maxSize: number, opts?: stream.TransformOptions) {
+    super({ ...opts, objectMode: true });
+  }
+
+  _transform(chunk: Data, _: BufferEncoding, callback: stream.TransformCallback) {
+    const chunkSize = JSON.stringify(chunk.json).length + chunk.pathname.length;
+    if (this.size + chunkSize > this.maxSize) {
+      this.push(this.batch);
+      this.batch = { json: {}, pathname: "" };
+      this.size = 0;
+    }
+    if (this.size === 0) {
+      this.batch = chunk;
+    } else {
+      const newPathname = this._findLongestCommonPrefix(chunk.pathname, this.batch.pathname);
+      const batchKey = this.batch.pathname.substring(newPathname.length + 1); // +1 to trim leading slash
+      const chunkKey = chunk.pathname.substring(newPathname.length + 1);
```

Comment:
The iterative approach would work. Not sure if `String.substring` trigger a copy. In that case, it would be O(n^2).

Might be more safe if we find the common prefix of `[]Data` in the end. Less objects allocated.

Reply:
Not sure I understand your comment. Do you mean you want to do all the batching, then find the prefix for each batch?

Reply:
Yeah. The chunk size can be a rough estimate. Doesn't need to be accurate.

I worry a bit that `substring` on each chunk will create a copy and make it O(n^2).
- Collect batches until the estimated size reach the payload size.
- Compact batches, find common prefix, send them out.

Reply:
Done.